crates/agent/src/publications/finalize.rs

```diff
@@ -1,12 +1,53 @@
+use super::UncommittedBuild;
 use std::collections::BTreeSet;
 
 use models::ModelDef;
 use tables::{BuiltRow, Validations};
 
+/// Inspect, and potentially modify a completed build. If the build has any errors after this
+/// function returns, then it will be considered a failed build.
+///
+/// Note that modifications to the build are _not_ persisted in the build database, and thus
+/// for the time being it would be incorrect to modify built specs here. It is permissible to
+/// remove built specs, though, since there's no harm in having extra specs in the build db.
+/// This restriction may be lifted in the future if we stop using sqlite databases, or if we
+/// refactor builds to apply `FinalizeBuild`s before persisting the sqlite database. For now,
+/// it does not seem worth the effort, though.
+pub trait FinalizeBuild {
+    fn finalize(&self, build: &mut UncommittedBuild) -> anyhow::Result<()>;
+}
+
+/// A `FinalizeBuild` that doesn't modify the build in any way.
+pub struct NoopFinalize;
+impl FinalizeBuild for NoopFinalize {
+    fn finalize(&self, _build: &mut UncommittedBuild) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+/// A `FinalizeBuild` that removes and built collection specs that are not referenced by other
```

Comment:
supernit: removes and built => removes any built

> Note that this must be used in conjuction with the `ExpandDraft`  `Initialize` impl, or else it can prune collections that are referenced by live specs that are not included in the draft.

Also, is it a problem that it might remove collections referenced only by pass-through live specs? I'm not seeing why.

Reply:
Yeah, this comment was based on the faulty assumption that we might prune draft specs that already have corresponding live specs. But we only consider _new_ specs for pruning, so there's no chance that a pruned draft spec might have been referenced by an existing live spec. I updated the comment.